File: russian-doll-envelopes/russianDollEnvelopes.java
public class Solution {
    public int maxEnvelopes(int[][] envelopes) {
        if (envelopes == null || envelopes.length == 0 || envelopes[0] == null || envelopes[0].length == 0) {
            return 0;
        }
        Arrays.sort(envelopes, new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                if (a[0] == b[0]) {
                    return b[1] - a[1];
                } else {
                    return a[0] - b[0];
                }
            } 
        });
        
        int size = 0;
        int[] dp = new int[envelopes.length];
        for (int[] x : envelopes) {
            int left = 0, right = size;
            int index = Arrays.binarySearch(dp, 0, size, x[1]);
            if (index < 0) {
                index = -(index + 1);
            }
            dp[index] = x[1];
            if (index == size) {
                size++;
            }
        }
        return size;
    }
}

File: russian-doll-envelopes/russianDollEnvelopes.py
class Solution(object):
    def myCmp(self, x, y):
        if x[0] == y[0]:
            return y[1] - x[1]
        else:
            return x[0] - y[0]
            
    def maxEnvelopes(self, envelopes):
        """
        :type envelopes: List[List[int]]
        :rtype: int
        """
        if envelopes == None or len(envelopes) == 0 or envelopes[0] == None or len(envelopes[0]) == 0:
            return 0
        
        envelopes.sort(cmp=self.myCmp)
        
        size = 0
        dp = [0] * len(envelopes)
        for x in envelopes:
            left = 0
            right = size
            i = bisect.bisect_left(dp, x[1], 0, size)
            dp[i] = x[1]
            if i == size:
                size += 1
                
        return size